packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -320,3 +328,169 @@ export function removeComments(ast: AST): void {
   const nodesWithComments = ast.find(Node).filter(nodePath => !!nodePath.node.comments);
   nodesWithComments.forEach(nodePath => (nodePath.node.comments = null));
 }
+
+/**
+ * Determines from a given AST of a file, whether the file has a default export or not.
+ */
+export function hasDefaultExport(ast: AST): boolean {
+  const hasDefaultDeclaration = ast.find(ExportDefaultDeclaration).size() > 0;
+  if (hasDefaultDeclaration) {
+    return true;
+  }
+
+  const hasDefaultSpecifier = ast.find(ExportDefaultSpecifier).size() > 0;
+  if (hasDefaultSpecifier) {
+    return true;
+  }
+
+  const hasNamedDefaultExport = ast.find(ExportSpecifier).some(specifier => specifier.node.exported.name === 'default');
+  if (hasNamedDefaultExport) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Extracts all identifiers from a "RestElement" within a named export declaration statement ("export const val = ...").
+ * "RestElements" are things like "...value" inside a destructuring assignment.
+ *
+ * Example:
+ * "export const { ...restElement1, bar: { ...restElement2 } } = { foo: 1, bar: { baz: 2 } }" --> ["restElement1", "restElement2"]
+ */
+function getExportIdentifiersFromRestElement(restElement: jscsTypes.RestElement): string[] {
+  const identifiers: string[] = [];
+
+  if (Identifier.check(restElement.argument)) {
+    identifiers.push(restElement.argument.name);
+  }
+
+  return identifiers;
+}
+
+/**
+ * Extracts all identifiers from an ArrayPattern within a destructured named export declaration
+ * statement ("export const [val] = [1]").
```

Comment:
I think we should reserve `val` for actual values, not their names. Also, what you're doing here has nothing to do with exports, does it? Wouldn't it work in any destructuring `ArrayPattern`? (Same with the function description in `getExportIdentifiersFromObjectPattern`.)

```suggestion
 * Extracts all identifier names (`'constName'`) from an destructuring assignment's ArrayPattern (the `[constName]` in `const [constName] = [1]`).
```

Reply:
I applied your suggestion and cleaned up the description a bit: 0002873d52758bb79d0c5f9d4b5c8c55f63541ce

As for the exports thing: This was very intentional and I will comment somewhere else in more detail on why I did it. (In short, the array patterns within export statements have less valid patterns than let's say on the right-hand side of an assignment. ;aking this function generic would have to make it handle way more edge cases)

Reply:
> the array patterns within export statements have less valid patterns than let's say on the right-hand side of an assignment

Yeah, but then you don't have `ObjectPattern`s and `ArrayPattern`s, you have `ObjectExpression`s and `ArrayExpressions`. I claim that the pattern versions only appear on the lefthand side, which AFAIK can only happen when destructuring.

That said, I don't feel super strongly about this, especially if we're possibly going to try to supplant it with something simpler, so let's leave it for now.